### PR TITLE
Create definitions for promoted constants.

### PR DESCRIPTION
**Pull request by cjgillot**
The current handling of promoted carries an `Option<Promoted>` everywhere to disambiguate the normal item from the constant.

This PR replaces the promoted constants by new definitions with new `DefKind::Promoted`.
This method allows to unify borrow-checking with the code path for inline consts.

The main caveat is creating even more definitions that do not have any HIR associated, triggering new corner cases where `local_def_id_to_hir_id` panics.

### CODE REVIEW

File: compiler/rustc_middle/src/mir/interpret/mod.rs [hunk 1]
```diff
@@ -142,19 +141,11 @@ pub struct GlobalId<'tcx> {
     /// For a constant or static, the `Instance` of the item itself.
     /// For a promoted global, the `Instance` of the function they belong to.
     pub instance: ty::Instance<'tcx>,
-
-    /// The index for promoted globals within their function's `mir::Body`.
-    pub promoted: Option<mir::Promoted>,
 }
```
Comment (RalfJung) [inline compiler/rustc_middle/src/mir/interpret/mod.rs:144]: Now GlobalId is just a newtype for Instance... is it even still worth having that type?

File: compiler/rustc_middle/src/mir/mod.rs [hunk 2]
```diff
@@ -184,18 +184,15 @@ impl RuntimePhase {
 #[derive(HashStable, TyEncodable, TyDecodable, TypeFoldable, TypeVisitable)]
 pub struct MirSource<'tcx> {
     pub instance: InstanceDef<'tcx>,
-
-    /// If `Some`, this is a promoted rvalue within the parent function.
-    pub promoted: Option<Promoted>,
 }
```
Comment (RalfJung) [inline compiler/rustc_middle/src/mir/mod.rs:187]: I guess same question here, is this newtype still useful?

**Comment by Nilstrieb** [comment]
Does it make sense to remove `local_def_id_to_hir_id` and only keep the fallible variant afterwards? Having these panicking functions seems like a footguns.

**Comment by oli-obk** [comment]
Considering there are additional executions of queries like `mir_borrowck` now, could it be that we're borrowcking promoteds twice now? Once as part of their parent, and once because they're their own mir body?

**Comment by cjgillot** [comment]
We stop borrow-checking promoted as part of the enclosing body, so they keep being checked only once. They now work exactly like inline consts.

However, we now run the rest of borrow-checking on them, and pay the extra cost of setting-up and cleaning-up the borrow-checking environment: renumbering regions, solving them, and wrapping that up into a nice `BorrowCheckResult`.

Profiling that led me to https://github.com/rust-lang/rust/pull/111753 and https://github.com/rust-lang/rust/pull/111759. This PR's perf will have to be re-checked on those land.

**Comment by cjgillot** [comment]
The performance degradation in incremental is due to a current limitation of creating new definitions. When we create a new definition, we make the creator query depend on the forever-red dep-node. This means that in the incremental scenario, we force `mir_promoted` to be recomputed if it ever promotes anything, and recursively this requires building MIR and THIR.

There is also a perf degradation in non-incremental, inside borrowck. I haven't found its cause yet.

**Comment by Dylan-DPC** [comment]
Closing this as it was an experiment and has too many conflicts

**Comment by oli-obk** [comment]
Since I forgot to cross link it: I'm working on fixing the perf issues here via https://github.com/rust-lang/rust/pull/115613

**Pull request closed without merging**